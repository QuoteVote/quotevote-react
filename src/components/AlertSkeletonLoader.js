import GridList from '@material-ui/core/GridList'
import GridListTile from '@material-ui/core/GridListTile'
import Skeleton from '@material-ui/lab/Skeleton'
import React from 'react'
import PropTypes from 'prop-types'
import { Grid } from '@material-ui/core'
import withWidth from '@material-ui/core/withWidth'

function AlertSkeletonLoader({ cols, width }) {
  const iterator = cols > 1 ? 12 : 3
  const numberOfRows = Array.from(Array(iterator).keys())
  if (cols > 1) {
    return (
      <GridList cols={width === 'xs' ? 1 : cols}>
        {numberOfRows.map((item) => (
          <GridListTile
            key={item}
            cols={1}
            style={{ marginTop: width === 'xs' ? -25 : -50 }}
          >
            <Skeleton animation="wave" height={300} />
          </GridListTile>
        ))}
      </GridList>
    )
  }

  return (
    <Grid
      container
      direction="column"
      justify="flex-start"
      alignItems="stretch"
      spacing={0}
      style={{ marginTop: -50 }}
    >
      {numberOfRows.map((item) => (
        <Grid item key={item} style={{ marginBottom: -100 }}>
          <Skeleton animation="wave" height={300} />
        </Grid>
      ))}
    </Grid>
  )
}

AlertSkeletonLoader.propTypes = {
  cols: PropTypes.number,
  width: PropTypes.oneOf(['lg', 'md', 'sm', 'xl', 'xs']).isRequired,
}

export default withWidth()(AlertSkeletonLoader)
