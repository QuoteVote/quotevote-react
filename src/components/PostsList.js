import React from 'react'
import PropTypes from 'prop-types'
import Skeleton from '@material-ui/lab/Skeleton'
import { useDispatch, useSelector } from 'react-redux'
import { SET_HIDDEN_POSTS, SET_SNACKBAR } from 'store/ui'
import { useMutation } from '@apollo/react-hooks'
import { GET_TOP_POSTS } from 'graphql/query'
import { UPDATE_POST_BOOKMARK } from 'graphql/mutations'
import { getGridListCols, useWidth } from 'utils/display'
import GridList from '@material-ui/core/GridList'
import GridListTile from '@material-ui/core/GridListTile'
import PostCard from './PostCard'

export function AlertSkeletonLoader({ width }) {
  const rows = Array.from(Array(12).keys())
  return (
    <GridList cols={getGridListCols[width]}>
      {rows.map((item) => (
        <GridListTile key={item} cols={1}>
          <Skeleton animation="wave" height={128} />
        </GridListTile>
      ))}
    </GridList>
  )
}

export function LoadPostsList({ data, width }) {
  const dispatch = useDispatch()
  const user = useSelector((state) => state.user.data)
  const hiddenPosts = useSelector((state) => state.ui.hiddenPosts)
  const snackbar = useSelector((state) => state.ui.snackbar)
  const limit = 12 + hiddenPosts.length
  const [updatePostBookmark, { error }] = useMutation(UPDATE_POST_BOOKMARK, {
    refetchQueries: [
      {
        query: GET_TOP_POSTS,
        variables: { limit, offset: 0, searchKey: '' },
      },
    ],
  })

  // !snackbar.open prevent dispatching action again once snackbar is already opened
  if (error && !snackbar.open) {
    dispatch(SET_SNACKBAR({
      type: 'danger',
      message: `Updating bookmark error: ${error}`,
      open: true,
    }))
  }

  const handleBookmark = (postId) => {
    // eslint-disable-next-line no-underscore-dangle
    updatePostBookmark({
      variables: { postId, userId: user._id },
      update: (cache, { data: updatedBookmark }) => {
        if (updatedBookmark) {
          dispatch(
            SET_SNACKBAR({
              type: 'success',
              message: 'Updated Successfully',
              open: true,
            })
          )
        }
      },
    })
  }

  const handleHidePost = (post) => {
    dispatch(SET_HIDDEN_POSTS(post._id))
  }

  if (!data || data.posts === 0) {
    return (
      <div style={{ width: '90%', textAlign: 'center' }}>
        <span>No posts fetched.</span>
        <br></br>
      </div>
    )
  }

  const rankedPosts = data.posts
    .map((post, index) => ({ ...post, rank: index + 1 }))
    .filter((post) => !hiddenPosts.includes(post._id))

  return (
    <GridList cols={getGridListCols[width]}>
      {rankedPosts.map((prop, key) => (
        <GridListTile key={key} cols={1}>
          <PostCard
            {...prop}
            onHidePost={handleHidePost}
            user={user}
            onBookmark={handleBookmark}
          />
        </GridListTile>
      ))}
    </GridList>
  )
}

export default function PostList({ Data, loading, limit }) {
  const width = useWidth()
  if (loading) return <AlertSkeletonLoader limit={limit} width={width} />
  return <LoadPostsList width={width} data={Data} />
}

PostList.propTypes = {
  Data: PropTypes.object.isRequired,
  loading: PropTypes.bool.isRequired,
  limit: PropTypes.number.isRequired,
}

AlertSkeletonLoader.propTypes = {
  width: PropTypes.object.isRequired,
}

LoadPostsList.propTypes = {
  width: PropTypes.object.isRequired,
  data: PropTypes.object.isRequired,
}
