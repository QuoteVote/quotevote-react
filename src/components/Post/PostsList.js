import React from 'react'
import PropTypes from 'prop-types'
import { useDispatch, useSelector } from 'react-redux'
import { SET_HIDDEN_POSTS } from 'store/ui'
import InfiniteScroll from 'react-infinite-scroller'
import { Grid } from '@material-ui/core'
import PostCard from './PostCard'
import AlertSkeletonLoader from '../AlertSkeletonLoader'
import LoadingSpinner from '../LoadingSpinner'

export function LoadPostsList({
  data, onLoadMore,
}) {
  const dispatch = useDispatch()
  const user = useSelector((state) => state.user.data)
  const hiddenPosts = useSelector((state) => state.ui.hiddenPosts) || []
  const handleHidePost = (post) => {
    dispatch(SET_HIDDEN_POSTS(post._id))
  }

  if (!data || data.posts === 0) {
    return (
      <div style={{ width: '90%', textAlign: 'center' }}>
        <span>No posts fetched.</span>
        <br />
      </div>
    )
  }

  const rankedPosts = data.posts.entities
    .map((post, index) => ({ ...post, rank: index + 1 }))
    .filter((post) => !hiddenPosts.includes(post._id))

  console.log(rankedPosts)

  const reordered = _.orderBy(rankedPosts, 'created', 'desc')

  console.log(reordered)

  const hasMore = data.posts.pagination.total_count > rankedPosts.length
  return (

    <InfiniteScroll
      pageStart={0}
      loadMore={onLoadMore}
      hasMore={hasMore}
      loader={<div className="loader" key={0}><LoadingSpinner size={30} /></div>}
    >
      <Grid
        container
        direction="column"
        justify="center"
        alignItems="stretch"
        spacing={0}
      >
        {rankedPosts.map((prop) => (
          <Grid item style={{ marginBottom: -25 }} key={prop._id}>
            <PostCard
              {...prop}
              onHidePost={handleHidePost}
              user={user}
              key={prop._id}
            />
          </Grid>
        ))}
      </Grid>
    </InfiniteScroll>
  )
}

LoadPostsList.propTypes = {
  data: PropTypes.object,
  onLoadMore: PropTypes.func.isRequired,
}

export default function PostList({
  data, loading, limit, fetchMore, variables, cols,
}) {
  if (loading) return <AlertSkeletonLoader limit={limit} rows={cols} />

  const newOffset = data && data.posts.entities.length
  return (
    <LoadPostsList
      data={data}
      onLoadMore={() => fetchMore({
        variables: {
          ...variables,
          offset: newOffset,
        },
        updateQuery: (prev, { fetchMoreResult }) => {
          if (!fetchMoreResult) return prev
          return {
            ...prev,
            posts: {
              ...fetchMoreResult.posts,
              entities: [
                ...prev.posts.entities,
                ...fetchMoreResult.posts.entities,
              ],
            },
          }
        },
      })}
    />
  )
}

PostList.propTypes = {
  data: PropTypes.object,
  loading: PropTypes.bool.isRequired,
  limit: PropTypes.number.isRequired,
  fetchMore: PropTypes.func.isRequired,
  variables: PropTypes.object.isRequired,
  cols: PropTypes.number.isRequired,
}
