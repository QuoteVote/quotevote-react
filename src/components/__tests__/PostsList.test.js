/* eslint-disable react/jsx-no-undef */
import React from 'react'
import PostsList, { LoadPostsList } from '../PostsList'
import AlertSkeletonLoader from '../AlertSkeletonLoader'

describe('PostsList component unit test', () => {
  let component
  const props = {
    Data: {
      posts: [],
    },
    loading: false,
    limit: 1,
  }

  beforeEach(async () => {
    component = shallow(
      <PostsList {...props} />
    )
  })

  it('renders PostsList without crashing', () => {
    expect(component).toMatchSnapshot()
  })

  it('should LoadPostsList if loading = false', async () => {
    const fnd = component.find(LoadPostsList)
    expect(fnd.exists()).toBeTruthy()
  })

  it('should exist AlertSkeletonLoader if loading = true', async () => {
    component.setProps({ loading: true })
    const fnd = component.find(AlertSkeletonLoader)
    expect(fnd.exists()).toBeTruthy()
  })
})
