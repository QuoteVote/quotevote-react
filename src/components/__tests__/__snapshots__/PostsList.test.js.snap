// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList component unit test renders PostsList without crashing 1`] = `ShallowWrapper {}`;
